libsoftwarecontainer/src/gateway/devicenode/devicenodelogic.cpp

```diff
@@ -18,33 +18,40 @@
  */
 
 #include "devicenodelogic.h"
+#include "devicenodegateway.h"
 
 namespace softwarecontainer {
 
-std::vector<DeviceNodeParser::Device>::iterator
+std::vector<std::shared_ptr<Device>>::iterator
 DeviceNodeLogic::findDeviceByName(const std::string name)
 {
-    std::vector<DeviceNodeParser::Device>::iterator item =
+    std::vector<std::shared_ptr<Device>>::iterator item =
             std::find_if(m_devList.begin(), m_devList.end(),
-            [&] (DeviceNodeParser::Device const &d) { return d.name == name; });
+            [&] (std::shared_ptr<Device> const &d) { return d->name == name; });
 
     return item;
 }
 
-bool DeviceNodeLogic::updateDeviceList(DeviceNodeParser::Device dev)
+bool DeviceNodeLogic::updateDeviceList(Device &dev)
 {
     auto item = findDeviceByName(dev.name);
 
     if (item == std::end(m_devList)) {
-        m_devList.push_back(dev);
+        std::shared_ptr<Device> newDevice(new Device(dev));
+        newDevice->isConfigured = false;
+        m_devList.push_back(std::move(newDevice));
     } else {
-        item->mode = calculateDeviceMode(item->mode, dev.mode);
+        auto mode = calculateDeviceMode(((*item))->mode, dev.mode);
+        if (mode != (*item)->mode) {
+            (*item)->isConfigured = false;
+            (*item)->mode = mode;
+        }
```

Comment:
If you return the shared_ptr directly, from findDeviceByName, this code would look way better.